File: tests/st/graph_kernel/expander/test_mint_binary_cross_entropy.py
# Copyright 2024 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

import os
import numpy as np
import mindspore.context as context
import mindspore
from mindspore import Tensor, mint
from mindspore.nn import Cell


class Net(Cell):
    def __init__(self):
        super(Net, self).__init__()
        self.binary_cross_entropy = mint.nn.functional.binary_cross_entropy

    def construct(self, x1, x2, weight, mode):
        return self.binary_cross_entropy(x1, x2, weight, mode)


def get_output(x1, x2, weight, mode, enable_graph_kernel=False):
    if enable_graph_kernel:
        context.set_context(jit_level='O1', graph_kernel_flags="--dump_as_text --enable_expand_ops=BinaryCrossEntropy")
    else:
        context.set_context(jit_level='O0')
    net = Net()
    output = net(x1, x2, weight, mode)
    output = output.asnumpy()
    if enable_graph_kernel:
        context.set_context(graph_kernel_flags="")
    return output


def run_basic(dtype, mode='mean', compare_precision=1e-4):
    def _remove_file(file_path):
        if os.path.exists(file_path):
            try:
                os.remove(file_path)
            except OSError:
                pass

    x1 = Tensor(np.random.normal(
        0.5, 0.01, [256, 256]).astype(np.float32), dtype=dtype)
    x2 = Tensor(np.random.normal(
        0.5, 0.01, [256, 256]).astype(np.float32), dtype=dtype)
    weight = Tensor(np.random.normal(
        0.5, 0.01, [256, 256]).astype(np.float32), dtype=dtype)
    expect = get_output(x1, x2, weight, mode, False)
    output = get_output(x1, x2, weight, mode, True)
    dump_file = "./graph_kernel_dump/dvm_kernel_{}.txt".format(os.getpid())
    try:
        np.testing.assert_allclose(expect, output, compare_precision, compare_precision)
    except Exception as ex:
        if os.path.isfile(dump_file):
            print("dump_file", dump_file)
            with open(dump_file, 'r') as f:
                for line in f:
                    print(line)
            _remove_file(dump_file)
        raise RuntimeError("Precision compare failed!\n{}".format(ex))
    _remove_file(dump_file)


def test_basic_ascend_f16():
    """
    Feature: test graph kernel mint.binary_cross_entropy
    Description: run test case on Ascend
    Expectation: the result match with expect
    """
    context.set_context(mode=context.GRAPH_MODE)
    run_basic(mindspore.float16, compare_precision=1e-3)


def test_basic_ascend_f32():
    """
    Feature: test graph kernel mint.binary_cross_entropy
    Description: run test case on Ascend
    Expectation: the result match with expect
    """
    context.set_context(mode=context.GRAPH_MODE)
    run_basic(mindspore.float32)


def test_basic_ascend_f32_none():
    """
    Feature: test graph kernel mint.binary_cross_entropy
    Description: run test case on Ascend
    Expectation: the result match with expect
    """
    context.set_context(mode=context.GRAPH_MODE)
    run_basic(mindspore.float32, 'none')


def test_basic_ascend_f32_none_corner_case():
    """
    Feature: test graph kernel mint.binary_cross_entropy
    Description: run test case on Ascend
    Expectation: the result match with expect, in this case, the result should be [100. 100. -0.]
    """
    context.set_context(mode=context.GRAPH_MODE)
    logits = Tensor(np.array([1, 0, 1]).astype(np.float32), dtype=mindspore.float32)
    labels = Tensor(np.array([0, 1, 1]).astype(np.float32), dtype=mindspore.float32)
    weight = Tensor(np.array([1, 1, 1]).astype(np.float32), dtype=mindspore.float32)
    expect = get_output(logits, labels, weight, 'none', False)
    output = get_output(logits, labels, weight, 'none', True)
    assert np.allclose(expect, output, 0.0001, 0.0001, equal_nan=True)


def test_basic_ascend_f32_sum():
    """
    Feature: test graph kernel mint.binary_cross_entropy
    Description: run test case on Ascend
    Expectation: the result match with expect
    """
    context.set_context(mode=context.GRAPH_MODE)
    run_basic(mindspore.float32, 'sum')

File: tests/ut/cpp/backend/ms_backend/graph_fusion/expander/test_mint_binary_cross_entropy.cc
/**
 * Copyright 2024 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <map>
#include <string>
#include <iostream>
#include "backend/ms_backend/graph_fusion/common/graph_kernel_common_test_suite.h"
#include "utils/anf_utils.h"
#include "abstract/abstract_value.h"
#include "common/graph_optimizer_test_framework.h"
#include "include/runtime/hardware_abstract/kernel_base/graph_fusion/graph_kernel_flags.h"
#include "backend/ms_backend/graph_fusion/adapter/graph_kernel_expander_cloud.h"
#include "include/backend/anf_runtime_algorithm.h"
#include "pre_activate/common/pattern_to_pattern_pass_utils.h"
#include "backend/ms_backend/graph_fusion/expander/base.h"
#include "ir/graph_utils.h"

namespace mindspore::graphkernel::test {
namespace {
struct BinaryCrossEntropyParams {
  ShapeVector logits;
  ShapeVector labels;
  ShapeVector weight;
  int64_t mode;  // 0-----ReduceSum, 1------ReduceMean, 2-------None
  ShapeVector expect_shape;
  TypePtr type;
};
}  // namespace

/// Feature: Test graph kernel BinaryCrossEntropy expander
/// Description: BinaryCrossEntropy will expanded
/// Expectation: After expand, the output shape and data type of sub graph should match expect
class TestBinaryCrossEntropyExpander : public TestGraphKernelExpander,
                                       public testing::WithParamInterface<BinaryCrossEntropyParams> {
  void SetUp() override {
    SetDeviceTarget(kAscendDevice);
    std::map<std::string, std::string> jit_config;
    jit_config["graph_kernel_flags"] = "--kernel_generator=DVM --enable_expand_ops=BinaryCrossEntropy";
    graphkernel::GraphKernelFlags::SaveJitConfig(jit_config);
  }
};

TEST_P(TestBinaryCrossEntropyExpander, BinaryCrossEntropy) {
  const auto &param = GetParam();
  ConstructGraph c;
  auto logits = c.NewTensorInput("logits", param.type, param.logits);
  auto labels = c.NewTensorInput("labels", param.type, param.labels);
  auto weight = c.NewTensorInput("weight", param.type, param.weight);
  auto mode = c.NewScalarInput("mode", MakeValue(param.mode), kInt64);
  auto op = c.NewCNodeWithBuildInfo("BinaryCrossEntropy", {logits, labels, weight, mode}, {});
  c.SetOutput(op);
  RunPass(c.GetGraph(), {std::make_shared<graphkernel::GraphKernelExpanderCloud>()});
  auto nodes = TopoSort(c.GetGraph()->get_return());
  for (const auto &node : nodes) {
    if (node != nullptr && AnfUtils::IsGraphKernel(node)) {
      CompareShapeAndType(node, 0, param.expect_shape, param.type->type_id());
    }
  }
  auto g = c.GetGraph();
  UT_CHECK_NULL(g);
  auto gknodes = GetAllGKNodes(g);
  EXPECT_EQ(gknodes.size(), 1);
}

INSTANTIATE_TEST_CASE_P(TestOpBinaryCrossEntropy, TestBinaryCrossEntropyExpander,
                        testing::Values(BinaryCrossEntropyParams{{16, 16}, {16, 16}, {16, 16}, 0, {}, kFloat16},
                                        BinaryCrossEntropyParams{{16, 16}, {16, 16}, {16, 16}, 0, {}, kFloat32},
                                        BinaryCrossEntropyParams{{16, 16}, {16, 16}, {16, 16}, 1, {}, kFloat16},
                                        BinaryCrossEntropyParams{{16, 16}, {16, 16}, {16, 16}, 1, {}, kFloat32},
                                        BinaryCrossEntropyParams{{16, 16}, {16, 16}, {16, 16}, 2, {16, 16}, kFloat16},
                                        BinaryCrossEntropyParams{{16, 16}, {16, 16}, {16, 16}, 2, {16, 16}, kFloat32}));
}  // namespace mindspore::graphkernel::test